### PR TITLE
sql/contention: add contention registry for global contention observability

### PR DESCRIPTION
**Pull request by asubiotto**
This PR adds a contention registry that groups roachpb.ContentionEvents by
tableID/indexID pair and collects metadata about them. This data structure is
backed by an LRU cache.

Release note: None

Addresses #57114

My goal here is to discuss the basic structure of the registry and get this basic version in. Once that's done, I'll work on:
1) Add a pointer to a global registry in the distsql receiver and add thread safety.
2) Adding a serialized version so registries can be sent between nodes and merged as well as sent back to the admin UI.
3) Implement a virtual table and status server endpoint that will merge and return these serialized versions.

cc @tbg

### CODE REVIEW

**Comment by asubiotto** [comment]
Friendly ping

**Review by RaduBerinde — APPROVED** [review]
Very nice code!

One thing is that we could make this more simple/efficient if we don't keep track of individual keys. If there's a chance we won't actually ever show the keys, we should decide now.. (CC @jordanlewis)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57736#-:-MOXOsd32HJ8DUPWXQq9:b-fkwjfu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @RaduBerinde)

___
*[pkg/sql/contention/registry.go, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57736#-MOXLkRO0aMCMdE83eJZ:-MOXLkRO0aMCMdE83eJ_:bevxdao) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c971edc791e932639818c9c48d9671c51d93add/pkg/sql/contention/registry.go#L89)):*
> ```Go
> 	// this index have spent contended.
> 	cumulativeContentionTime time.Duration
> 	// orderedKeyMap is an LRU cache that keeps track of up to
> ```

LRU is questionable here.. Should we keep the "top" txns (in terms of the total duration), rather than LRU? (can be a TODO)
___
*[pkg/sql/contention/registry.go, line 168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57736#-MOXONIh1cGka1TTDrdW:-MOXONIh1cGka1TTDrdX:b-urdusl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c971edc791e932639818c9c48d9671c51d93add/pkg/sql/contention/registry.go#L168)):*
> ```Go
> }
> 
> // Registry is an object that keeps track of aggregated contention information.
> ```

[nit] consider moving the struct to the top of the file, to serve as a kind of file-level doc.
___
*[pkg/sql/contention/registry.go, line 227 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57736#-MOXOfaY5JQQCv41rVoL:-MOXOfaY5JQQCv41rVoM:b-75v0xu) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c971edc791e932639818c9c48d9671c51d93add/pkg/sql/contention/registry.go#L227)):*
> ```Go
> 		writeChild(prefix, fmt.Sprintf("cumulative contention time: %s\n", v.cumulativeContentionTime))
> 		writeChild(prefix, fmt.Sprintf("keys:\n"))
> 		keyPrefix := prefix + "\t"
> ```

[nit] Tabs can lead to inconsistent alignment, why not just spaces?
___
*[pkg/sql/contention/testdata/contention_registry, line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57736#-MOXPE-3C0isFbyRf17R:-MOXPE-3C0isFbyRf17S:b2md6vm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c971edc791e932639818c9c48d9671c51d93add/pkg/sql/contention/testdata/contention_registry#L4)):*
> ```
> ----
> 
> evcheck tableid=1 indexid=1 key=key id=a duration=1
> ```

Very cool tests!

**Comment by tbg** [comment]
Nice! I only skimmed this but it's great to see this come "together" (KV is still going to hold up their end of the bargain but we're not quite there yet).

**Review by asubiotto — COMMENTED** [review]
TFTRs!

> One thing is that we could make this more simple/efficient if we don't keep track of individual keys. If there's a chance we won't actually ever show the keys, we should decide now..

Agreed cc @awoods187 @annebirzin , what do you think? Should we keep track of individual keys for contention or do we prefer to just show the index for privacy?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57736#-:-MOfDL2PDkQCuEEOvRw7:b-ob4iye)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @RaduBerinde)

___
*[pkg/sql/contention/registry.go, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57736#-MOXLkRO0aMCMdE83eJZ:-MOfCacmEcO-QztMjMLH:b-81w2li) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c971edc791e932639818c9c48d9671c51d93add/pkg/sql/contention/registry.go#L89)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

LRU is questionable here.. Should we keep the "top" txns (in terms of the total duration), rather than LRU? (can be a TODO)
</blockquote></details>

I chose LRU because I think it's more important to show a current contention view rather than the largest contention problems since contention began. My thinking is that a user cares a lot more about what the current largest hotspot is. If we were to evict based on the shortest duration txns, we could get in a state where we had an index with a ton of contention, the contention problem was fixed, but the index still remains in the contention registry and won't be evicted until relatively more contention happens on many other keys.
___
*[pkg/sql/contention/registry.go, line 227 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57736#-MOXOfaY5JQQCv41rVoL:-MOfBdQzDh-XbgKdoFKI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c971edc791e932639818c9c48d9671c51d93add/pkg/sql/contention/registry.go#L227)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Tabs can lead to inconsistent alignment, why not just spaces?
</blockquote></details>

Done.
___
*[pkg/sql/contention/testdata/contention_registry, line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57736#-MOXPE-3C0isFbyRf17R:-MOfBws96NszK6aGpWCS:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c971edc791e932639818c9c48d9671c51d93add/pkg/sql/contention/testdata/contention_registry#L4)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Very cool tests!
</blockquote></details>

Thanks!

**Comment by asubiotto** [comment]
One other question, what do we think about the constant registry size values (i.e. max indexes/keys to keep track of)?

**Review by RaduBerinde — COMMENTED** [review]
:lgtm:

The values seem reasonable. We will probably need to do some benchmarking at some point to understand impact.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57736#-:-MOn-qE34eQaJwfkcXK4:b-rr4qbo)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/contention/registry.go, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57736#-MOXLkRO0aMCMdE83eJZ:-MOgwOoz2AYtqVfXRqlz:b6i81t9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c971edc791e932639818c9c48d9671c51d93add/pkg/sql/contention/registry.go#L89)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I chose LRU because I think it's more important to show a current contention view rather than the largest contention problems since contention began. My thinking is that a user cares a lot more about what the current largest hotspot is. If we were to evict based on the shortest duration txns, we could get in a state where we had an index with a ton of contention, the contention problem was fixed, but the index still remains in the contention registry and won't be evicted until relatively more contention happens on many other keys.
</blockquote></details>

That makes sense, not sure what I was thinking (I might have thought about it as if it was a per-statement structure, but even there LRU makes more sense).

**Comment by asubiotto** [comment]
TFTRs!

Will merge to avoid having this PR sit around for too long. We can always come back and simplify this code if we decide not to track individual keys. It'll be easier to remove this code than add it back in if we need it.

bors r=RaduBerinde

**Pull request merged**
